qringbuf.go

```diff
@@ -1,3 +1,168 @@
+// Package qringbuf provides a concurrency-friendly zero-copy abstraction of
+// io.ReadAtLeast() over a pre-allocated ring-buffer, populated asynchronously
+// by a standalone goroutine. It is primarily designed for processing a series
+// of arbitrary streams each comprised of variable-length records.
+//
+// Specifically an object of this package makes the following guarantees:
+//  • Memory is allocated only at construction time, never during streaming
+//  • StartFill() spawns off a single goroutine (collector) which terminates when:
+//    ◦ It reaches readLimit, if one was supplied
+//    ◦ It receives any error from the wrapped reader (including io.EOF)
+//    ◦ It had a chance to evaluate a signal received from StopFill()
+//      (most io.Read() operations are not cancelable, and may block forever)
+//  • Every call to NextRegion() blocks until it can return:
+//    ◦ A *Region object representing a contiguous slice of at least MinRegion bytes
+//    ◦ A smaller "trailing" *Region and an io.EOF or ErrCollectorStopped
+//    ◦ A nil-Region and any other error
+//  • *Region.Bytes() is always a slice of the underlying buffer, no data copying takes place
+//  • Data backing a *Region is guaranteed to remain intact provided:
+//    ◦ NextRegion() has not been called again allowing writes into the buffer
+//    ◦ *Region.Reserve() was invoked, which blocks writes until a *Region.Release()
+//
+// Examples
+//
+// In code the basic usage looks roughly like this:
+//
+//  qrb, initErr := qringbuf.NewFromReader( *someReader, qringbuf.Config{ … } )
+//  …
+//  readLimit = int64( stopReadingAfterThatManyBytes )
+//  startErr := qrb.StartFill( readLimit )
+//  …
+//  var available, processed int
+//  for {
+//      reg, streamErr := qrb.NextRegion( available - processed )
```

Comment:
I'd have to think a bit about why the parameter to NextRegion is `available - processed`. Could you add a comment explaining this?

Reply:
Attempted to address in https://github.com/ipfs/go-qringbuf/pull/6/commits/fbfe0db2fda82 Let me know if not sufficiently clear...